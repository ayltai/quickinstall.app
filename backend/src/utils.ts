import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
