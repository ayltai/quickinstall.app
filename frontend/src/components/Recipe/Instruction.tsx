import Box from '@mui/material/Box';
import Link from '@mui/material/Link';
import Typography from '@mui/material/Typography';
import React, { type FC, } from 'react';
import { useTranslation, } from 'react-i18next';

import { useAppSelector, } from '../../hooks';
import type { Ingredient, } from '../../models';
import { Instruction, type InstructionProps, } from '../Instruction';

/**
 * An instruction dialog that informs users about how to install the selected packages.
 */
export const RecipeInstruction : FC<InstructionProps> = props => {
    const { transactionId, ...rest } = props;

    const { ingredients, } : { ingredients : Ingredient[],     } = useAppSelector(state => state.recipe);

    const { product, themeMode,   } : { product : 'win/packages' | 'mac/packages' | 'tweaks', themeMode   : 'light' | 'dark', } = useAppSelector(state => state.preference);

    const { t, } = useTranslation();

    return (
        <Instruction
            transactionId={transactionId}
            title={t(product === 'win/packages' ? 'packages.checkout.copy.win' : 'packages.checkout.copy.mac')}
            {...rest}>
            {ingredients.some(ingredient => ingredient.source === 'App Store') && (
                <>
                    <Typography>
                        {t('packages.checkout.instructions')}
                    </Typography>
                    <Link
                        color='primary'
                        href='https://support.apple.com/en-gb/guide/app-store/fir6253293d/mac'
                        target='_blank'
                        rel='noreferrer'>
                        https://support.apple.com/en-gb/guide/app-store/fir6253293d/mac
                    </Link>
                    <Box
                        sx={{
                            borderRadius    : '1rem',
                            backgroundColor : themeMode === 'light' ? '#dcedc8' : '#558b2f',
                        }}
                        padding={2}>
                        {ingredients.filter(ingredient => ingredient.source === 'App Store').map(ingredient => (
                            <Typography key={ingredient.id}>
                                {`â€¢ ${ingredient.name}`}
                            </Typography>
                        ))}
                    </Box>
                </>
            )}
        </Instruction>
    );
};
