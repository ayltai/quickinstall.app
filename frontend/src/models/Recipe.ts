import type { Ingredient, } from './Ingredient';

export type Recipe = {
    id           : string,
    creationDate : Date,
    ingredients  : Ingredient[],
};
