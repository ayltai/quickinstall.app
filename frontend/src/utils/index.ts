import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
