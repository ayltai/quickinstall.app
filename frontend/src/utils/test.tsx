import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
