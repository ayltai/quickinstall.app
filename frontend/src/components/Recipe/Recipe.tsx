import React, { useEffect, useState, } from 'react';

import { useAddRecipeMutation, } from '../../apis';
import { useAppDispatch, useAppSelector, } from '../../hooks';
import type { Ingredient, } from '../../models';
import { removeIngredient, } from '../../states';
import { handleError, } from '../../utils';
import { DetailsView, } from '../DetailsView';
import { ShoppingCartDetail, } from '../ShoppingCartDetail';
import { RecipeInstruction, } from './Instruction';

/**
 * A recipe contains a list of selected ingredients.
 */
export const Recipe = () => {
    const [ openInstruction, setOpenInstruction, ] = useState(false);
    const [ selected,        setSelected,        ] = useState<Ingredient | undefined>();
    const [ recipeId,        setRecipeId,        ] = useState<string>();

    const { product, } : { product : 'win/packages' | 'mac/packages' | 'tweaks', } = useAppSelector(state => state.preference);

    const { ingredients, } : { ingredients : Ingredient[], } = useAppSelector(state => state.recipe);

    const [ addRecipe, { data, error, reset, }, ] = useAddRecipeMutation();

    const dispatch = useAppDispatch();

    const handleSelect = (item : Ingredient) => setSelected(item);

    const handleDelete = (item : Ingredient) => dispatch(removeIngredient(item));

    const handleNext = () => {
        setOpenInstruction(true);

        addRecipe({
            platform : product.split('/')[0] as 'win' | 'mac',
            items    : ingredients.filter((ingredient : Ingredient) => product === 'win/packages' ? ingredient.source === 'Chocolatey' : ingredient.source === 'Homebrew' || ingredient.source === 'Homebrew (Cask)' || ingredient.source === 'App Store'),
        }).catch(handleError);
    };

    const handleCloseInstruction = () => {
        setOpenInstruction(false);

        reset();
    };

    useEffect(() => {
        if (data) setRecipeId(data);
    }, [ data, ]);

    useEffect(() => {
        if (error) handleError(error);
    }, [ error, ]);

    return (
        <ShoppingCartDetail
            items={ingredients.filter((ingredient : Ingredient) => product === 'win/packages' ? ingredient.source === 'Chocolatey' : ingredient.source === 'Homebrew' || ingredient.source === 'Homebrew (Cask)' || ingredient.source === 'App Store')}
            detailsView={
                <DetailsView
                    color={product === 'win/packages' ? 'info' : 'primary'}
                    description={selected?.description}
                    infoUrl={selected?.infoUrl}
                    source={selected?.source ?? ''} />
            }
            instruction={recipeId ? (
                <RecipeInstruction
                    open={openInstruction}
                    color={product === 'win/packages' ? 'info' : 'primary'}
                    transactionId={recipeId}
                    onClose={handleCloseInstruction} />
            ) : undefined}
            instructionShown={openInstruction}
            onSelect={handleSelect}
            onDelete={handleDelete}
            onNext={handleNext} />
    );
};
