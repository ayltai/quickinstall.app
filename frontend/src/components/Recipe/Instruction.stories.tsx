import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';
import { Provider, } from 'react-redux';

import { createStore, } from '../../utils/test';
import { RecipeInstruction, } from './Instruction';

const meta : Meta<typeof RecipeInstruction> = {
    component : RecipeInstruction,
    title     : 'Components/Data Presentations/Recipe Instruction',
    tags      : [
        'autodocs',
    ],
    argTypes  : {
        transactionId : {
            description : 'The ID of the transaction',
        },
    },
};

export default meta;

type Story = StoryObj<typeof RecipeInstruction>;

const store = createStore({
    recipe : {
        ingredients : [
            {
                id          : '1',
                name        : 'spotify',
                description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                author      : 'Spotify',
                source      : 'App Store',
            },
            {
                id          : '2',
                name        : 'Spotify Player',
                description : 'Aenean lacinia bibendum nulla sed consectetur. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Donec sed odio dui. Donec ullamcorper nulla non metus auctor fringilla.',
                author      : 'Spotify',
                source      : 'App Store',
            },
        ],
    },
});

export const Default : Story = {
    args : {
        open          : true,
        transactionId : '1',
    },
    decorators : [
        StoryComponent => (
            <Provider store={store}>
                <StoryComponent />
            </Provider>
        ),
    ],
};
