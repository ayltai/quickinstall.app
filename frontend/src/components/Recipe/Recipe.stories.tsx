import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';
import { Provider, } from 'react-redux';

import { createStore, } from '../../utils/test';
import { Recipe, } from './Recipe';

const meta : Meta<typeof Recipe> = {
    component : Recipe,
    title     : 'Components/Data Presentations/Recipe',
    tags      : [
        'autodocs',
    ],
    argTypes  : {
    },
};

export default meta;

type Story = StoryObj<typeof Recipe>;

const store = createStore({
    preference : {
        product : 'mac/packages',
    },
    recipe     : {
        ingredients : [
            {
                id          : '1',
                name        : 'spotify',
                description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                author      : 'Homebrew',
                source      : 'Homebrew',
            },
            {
                id          : '2',
                name        : 'Spotify Player',
                description : 'Aenean lacinia bibendum nulla sed consectetur. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Donec sed odio dui. Donec ullamcorper nulla non metus auctor fringilla.',
                author      : 'Homebrew',
                source      : 'Homebrew (Cask)',
            },
        ],
    },
});

export const Default : Story = {
    args       : {
    },
    decorators : [
        StoryComponent => (
            <Provider store={store}>
                <StoryComponent />
            </Provider>
        ),
    ],
};
