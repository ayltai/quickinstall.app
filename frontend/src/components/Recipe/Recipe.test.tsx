import React from 'react';

import { render, } from '../../utils/test';
import { Recipe, } from './Recipe';

describe('<Recipe />', () => {
    it('renders correctly', () => expect(render(
        <Recipe />,
        {
            recipe : {
                ingredients : [
                    {
                        id          : '1',
                        name        : 'spotify',
                        description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                        author      : 'Homebrew',
                        source      : 'Homebrew',
                    },
                    {
                        id          : '2',
                        name        : 'Spotify Player',
                        description : 'Aenean lacinia bibendum nulla sed consectetur. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Donec sed odio dui. Donec ullamcorper nulla non metus auctor fringilla.',
                        author      : 'Homebrew',
                        source      : 'Homebrew (Cask)',
                    },
                ],
            },
        }
    )).toMatchSnapshot());
});
