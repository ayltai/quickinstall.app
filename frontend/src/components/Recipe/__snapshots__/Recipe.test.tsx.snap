// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Recipe /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-30re44-MuiStack-root"
      >
        <div
          class="MuiStack-root css-1sf5043-MuiStack-root"
        >
          <div
            class="MuiBox-root css-i9gxme"
          />
          <div
            class="MuiStack-root css-hy5tm-MuiStack-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ykxrwf-MuiSvgIcon-root"
              data-testid="ErrorOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8"
              />
            </svg>
            <h5
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1b7t60j-MuiTypography-root"
            >
              cart.empty
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
            >
              cart.hint
            </p>
          </div>
          <div
            class="MuiBox-root css-i9gxme"
          />
        </div>
        <div
          aria-hidden="true"
          class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
          style="opacity: 0; visibility: hidden;"
        >
          <div
            class="MuiStack-root css-6rg9iv-MuiStack-root"
          >
            <span
              class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-mod6k9-MuiCircularProgress-root"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </span>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1r0phtf-MuiTypography-root"
            >
              cart.processing
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-30re44-MuiStack-root"
    >
      <div
        class="MuiStack-root css-1sf5043-MuiStack-root"
      >
        <div
          class="MuiBox-root css-i9gxme"
        />
        <div
          class="MuiStack-root css-hy5tm-MuiStack-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ykxrwf-MuiSvgIcon-root"
            data-testid="ErrorOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8"
            />
          </svg>
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1b7t60j-MuiTypography-root"
          >
            cart.empty
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            cart.hint
          </p>
        </div>
        <div
          class="MuiBox-root css-i9gxme"
        />
      </div>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
        style="opacity: 0; visibility: hidden;"
      >
        <div
          class="MuiStack-root css-6rg9iv-MuiStack-root"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-mod6k9-MuiCircularProgress-root"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1r0phtf-MuiTypography-root"
          >
            cart.processing
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
